zap/src/main/java/org/zaproxy/zap/session/ScriptBasedSessionManagementMethodType.java

```diff
@@ -0,0 +1,752 @@
+import org.zaproxy.zap.view.LayoutHelper;
+
+/**
+ * The type corresponding to a {@link SessionManagementMethod} for web applications that use Session
+ * Management Scripts, which can hopefully model any custom behaviour required
+ *
+ * @since TODO add version
+ */
+public class ScriptBasedSessionManagementMethodType extends SessionManagementMethodType {
+
+    public static final String CONTEXT_CONFIG_SESSION_MGMT_SCRIPT =
+            ExtensionSessionManagement.CONTEXT_CONFIG_SESSION + ".script";
+    public static final String CONTEXT_CONFIG_SESSION_MGMT_SCRIPT_NAME =
+            CONTEXT_CONFIG_SESSION_MGMT_SCRIPT + ".name";
+    public static final String CONTEXT_CONFIG_SESSION_MGMT_SCRIPT_PARAMS =
+            CONTEXT_CONFIG_SESSION_MGMT_SCRIPT + ".params";
+
+    private static final int METHOD_IDENTIFIER = 2;
+
+    private static final Logger LOG =
+            Logger.getLogger(ScriptBasedSessionManagementMethodType.class);
+
+    private static final String METHOD_NAME =
+            Constant.messages.getString("sessionmanagement.method.sc.name");
+
+    private static final String API_METHOD_NAME = "scriptBasedSessionManagement";
+
+    public static final String SCRIPT_TYPE_SESSION = "session";
+
+    private static final ImageIcon SCRIPT_ICON_SESSION =
+            new ImageIcon(
+                    ScriptBasedSessionManagementMethodType.class.getResource(
+                            "/resource/icon/16/script-session.png"));
+
+    private static ExtensionScript extensionScript;
+
+    public class ScriptBasedSessionManagementMethod implements SessionManagementMethod {
+
+        private ScriptWrapper script;
+        private SessionWrapper session;
+        private Map<String, String> paramValues;
+
+        /**
+         * Always return a new SessionWrapper, but reuse the session if one exists
+         *
+         * @param msg
+         * @return a new SessionWrapper
+         */
+        private SessionWrapper getSessionWrapper(HttpMessage msg) {
+            if (session != null) {
+                return new SessionWrapper(session.getSession(), msg, paramValues);
+            }
+            session = new SessionWrapper(new ScriptBasedSession(), msg, paramValues);
+            return session;
+        }
+
+        @Override
+        public boolean isConfigured() {
+            // Always configured
+            return true;
+        }
+
+        @Override
+        public SessionManagementMethodType getType() {
+            return new ScriptBasedSessionManagementMethodType();
+        }
+
+        @Override
+        public WebSession extractWebSession(HttpMessage msg) {
+            SessionScript sessionScript = getScriptInterface(script);
+            if (sessionScript != null) {
+                SessionWrapper wrapper = getSessionWrapper(msg);
+                try {
+                    sessionScript.extractWebSession(wrapper);
+                } catch (Exception e) {
+                    LOG.error(e.getMessage(), e);
```

Comment:
Could call `getScriptsExtension().handleScriptException(wrapper, e);` instead. There are other script calls that need the try/catch block.